mvc/models/model-binding.rst

```diff
@@ -1,8 +1,89 @@
+   :language: c#
+   :lines: 61-76  
+   
+.. note:: The strings in the URL route are not case sensitive.
+
+MVC checks for model binding attributes applied to the action method's parameters. If it finds a parameter attribute it will use the data source specified by the attribute. Otherwise, for any property that has a string converter, it looks for a value with a matching name. In the above example the only matching item is an integer named `id`, so then the value provider associates its values with action method parameters. Below is a list of the data sources in the order that model binding looks through them:
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method. (including jQuery POST requests).
+#. ``Route values``: The set of route values provided by `routing <https://docs.asp.net/projects/mvc/en/latest/controllers/routing.html>`_. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked for a key named ``id`` and there is nothing named ``id`` in the form values, it moved on to the route values looking for that key. In our example, it's a match. Binding happens, and the value is converted to the integer 2. The same request using Edit(string id) would convert to the string "2". 
+
+So far the example uses primitive types. If the action method's parameter were a complex type such as the ``Movie`` type, which contains both primitive and complex types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types looking for matches. Model binding looks for the pattern `parameter_name.property_name` to perform binding. If it doesn't find matching objects, it will attempt to bind using  just the property name. For complex types such as ``Collection`` types model binding looks for matches to `parameter_name[index]` or just `[index]`. Model binding treats  `Dictionary`` types similarly, asking for `parameter_name[key]` or just `[key]`, as long as they keys are simple types. 
+
+In order for binding to happen, members of complex model types must be public, writable properties containing a default public constructor. Public fields are not bound. If properties are arrays or IEnumerable types, the array properties or types in the IEnumerable must be writable. Model binding handles some objects differently than most. ``ICollection<T>`` is supported for read only properties.  
+
+When a parameter is bound, model binding stops looking for parameters with that name and it moves on to bind the next parameter. If binding fails, MVC does not throw an error. The error is recorded in the ``ModelState.Errors`` collection, and model binding attempts to bind to the next parameter or property. You can query for model state errors by checking the ``ModelState.IsValid`` method. 
```

Comment:
Suggest focusing on just `ModelState.IsValid` here. The next layer down (`ModelState.ValidationState`, `ModelState.ErrorCount`, indexing into the `ModelState`, et cetera) is used mostly in debug scenarios, when writing a custom model binder, or when writing an action aware of common errors and has some way to handle them. In the second case, the action might check `ModelState[key].Errors` or `ModelState[key].ValidationState` before taking action.

###### Details

This paragraph and the note below confuse the `ModelState` property on the `Controller` type with a `class` that used to exist in MVC named `ModelState` (renamed in late October to `ModelStateEntry`). The `ModelState` property is of type `ModelStateDictionary`. It is a dictionary of `ModelStateEntry`s. `ModelStateEntry` is where the `Errors` property lives. So,

* `ModelState.Errors` -> `ModelStateEntry.Errors` or perhaps `ModelState[key].Errors`.
* `ModelState.IsValid` is correct in any `Controller` subclass.
* `ModelState.Errors` collection is a property of the controller's `ModelStateDictionary` object -> [some simpler version of...] each entry in the controller's `ModelState` property (which has type `ModelStateDictionary`) is a `ModelStateEntry` and `ModelStateEntry` has an `Errors` property.
* if there are items in this collection -> if any entry in the `ModelStateDictionary` has a non-empty `Errors` collection

Also the `ModelState.ValidationState` property is what _really_ controls `ModelState.IsValid`. It rolls up the `ModelStateEntry.ValidationState` of all entries in the dictionary. The `ModelStateDictionary` is invalid if any entry is invalid _or_ not validated. So there may be nothing in any `ModelStateEntry.Errors` collection when `ModelState.IsValid` is `false`.